src/bank.rs

```diff
@@ -936,6 +959,54 @@ impl Bank {
     pub fn set_finality(&self, finality: usize) {
         self.finality_time.store(finality, Ordering::Relaxed);
     }
+
+    pub fn store_account_subscription(&self, pubkey: Pubkey, sink: Sink<Account>) {
+        let mut subscriptions = self.account_subscriptions.write().unwrap();
+        subscriptions.insert(pubkey, sink);
+    }
+
+    pub fn remove_account_subscrtion(&self, pubkey: Pubkey) -> bool {
+        let mut subscriptions = self.account_subscriptions.write().unwrap();
+        if subscriptions.contains_key(&pubkey) {
+            subscriptions.remove(&pubkey);
+            true
+        } else {
+            false
+        }
+    }
+
+    fn check_account_subscriptions(&self, pubkey: &Pubkey, account: &Account) {
+        if let Some(sink) = self.account_subscriptions.write().unwrap().get(pubkey) {
+            sink.notify(Ok(account.clone())).wait().unwrap();
+        }
+    }
+
+    pub fn store_signature_subscription(
+        &self,
+        signature: Signature,
+        sink: Sink<RpcSignatureStatus>,
+    ) {
+        let mut subscriptions = self.signature_subscriptions.write().unwrap();
+        subscriptions.insert(signature, sink);
+    }
+
+    pub fn remove_signature_subscription(&self, signature: Signature) -> bool {
+        let mut subscriptions = self.signature_subscriptions.write().unwrap();
+        if subscriptions.contains_key(&signature) {
+            subscriptions.remove(&signature);
+            true
+        } else {
+            false
+        }
+    }
+
+    fn check_signature_subscriptions(&self, signature: &Signature) {
+        if let Some(sink) = self.signature_subscriptions.write().unwrap().get(signature) {
+            sink.notify(Ok(RpcSignatureStatus::Confirmed))
```

Comment:
Automatically unsubscribing a signature subscription on confirmation seems nice for the client, as there's only ever a single notification for a given signature.

Do we notify on failure too (such as ProgramFailure)?

Reply:
Nice, I like unsubscribe on confirmation.
We don't currently notify on failure, but I can piece that in. You want it?

Reply:
Yes please.  Let's surface all final errors (ie, everything but `SignatureNotFound`), to help the client move along faster than waiting for an eventual timeout due to no successful confirmation.